### PR TITLE
[b-7] currency conversion

### PR DESCRIPTION
**Pull request by Jokeroff**

### CODE REVIEW

File: src/main/java/com/lebediev/movieland/service/conversion/CachedCurrency.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+            throw new RuntimeException("Error in connection for url = " + url + " : " + e);
+        }
+    }
+    public static List<CurrencyEntity> getCurrencyList(String json) {
+        log.info("Start getting currency list from json");
+        long startTime = System.currentTimeMillis();
+        List<CurrencyEntity> currencyEntityList = new ArrayList<>();
+        try {
+            ObjectMapper objectMapper = new ObjectMapper();
+            JsonNode rootNode = objectMapper.readTree(json);
```
Comment (Ramoid) [inline src/main/java/com/lebediev/movieland/service/conversion/CachedCurrency.java]: можно вместо получения стринга в getCurrenciesFromUrl через цепочку  BufferedReader  InputStreamReader напрямую заюзать
objectMapper.readTree(InputStream)

File: src/main/java/com/lebediev/movieland/service/conversion/CurrencyConverter.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    private final static Logger log = LoggerFactory.getLogger(CurrencyConverter.class);
+    @Autowired
+    private CachedCurrency cachedCurrency;
+
+    public Movie convertPrice(Movie movie, Currency currency){
+        log.info("Start converting price for movie with id: {} for currency: {}", movie.getMovieId(), currency);
+        long startTime = System.currentTimeMillis();
+        CurrencyEntity currencyEntity = cachedCurrency.getCurrencyEntity(currency);
+        double rate = currencyEntity.getRate();
+        movie.setPrice(movie.getPrice()/rate);
```
Comment (Ramoid) [inline src/main/java/com/lebediev/movieland/service/conversion/CurrencyConverter.java]: Округление с BigDecimal

Comment (Jokeroff) [reply src/main/java/com/lebediev/movieland/service/conversion/CurrencyConverter.java]: done, thank you!

File: src/main/java/com/lebediev/movieland/service/conversion/CachedCurrency.java (hunk 1 above)
Comment (Jokeroff) [reply src/main/java/com/lebediev/movieland/service/conversion/CachedCurrency.java]: не знал, спасибо
готово

File: src/main/java/com/lebediev/movieland/service/conversion/CachedCurrency.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+            reader.close();
+            log.info("Finish getting currency list from url: {}. It took {} ms",url, System.currentTimeMillis() - startTime);
+            return currencyEntityList;
+
+        }  catch (IOException e) {
+            throw new RuntimeException("Error in connection for url = " + url + " : " + e);
+        }
+    }
+
+    public static boolean isExpired(CurrencyEntity currencyEntity){
```
Comment (juliamanunia) [inline src/main/java/com/lebediev/movieland/service/conversion/CachedCurrency.java]: url returns CCYs for current day, do you really want to check it for expiration?

File: src/main/java/com/lebediev/movieland/service/conversion/CachedCurrency.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+            if(currencyEntity.getCurrency() == currency){
+                log.info("Finish getting currency from cache. It took {} ms", System.currentTimeMillis() - startTime);
+                return currencyEntity;
+            }
+        }
+        throw new RuntimeException("Could not get cached currency " + currency);
+    }
+
+
+    public static List<CurrencyEntity> getCurrenciesFromUrl(String url){
```
Comment (juliamanunia) [inline src/main/java/com/lebediev/movieland/service/conversion/CachedCurrency.java]: why public?

File: src/main/java/com/lebediev/movieland/service/conversion/CachedCurrency.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+import java.net.URLConnection;
+import java.security.Security;
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+import java.util.List;
+
+@Service
+public class CachedCurrency {
+    private final static Logger log = LoggerFactory.getLogger(CachedCurrency.class);
```
Comment (juliamanunia) [inline src/main/java/com/lebediev/movieland/service/conversion/CachedCurrency.java]: if static, should be upper case

File: src/main/java/com/lebediev/movieland/service/conversion/CachedCurrency.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    private volatile List<CurrencyEntity> currencyList;
+    private final static DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd.MM.yyyy");
```
Comment (juliamanunia) [inline src/main/java/com/lebediev/movieland/service/conversion/CachedCurrency.java]: upper case

File: src/main/java/com/lebediev/movieland/service/conversion/CachedCurrency.java (hunk 4 above)
Comment (Jokeroff) [reply src/main/java/com/lebediev/movieland/service/conversion/CachedCurrency.java]: thank you, changed. I had plans to test it somehow )

File: src/main/java/com/lebediev/movieland/service/conversion/CachedCurrency.java (hunk 3 above)
Comment (Jokeroff) [reply src/main/java/com/lebediev/movieland/service/conversion/CachedCurrency.java]: for case when we could not obtain actual data and our cache will contain expired rates

**Pull request merged**